pkg/services/statefetcher/statefetcher.go

```diff
@@ -0,0 +1,544 @@
+	}
+	startIndex := uint32(1)
+	for {
+		prm := client.PrmObjectSearch{}
+		filters := object.NewSearchFilters()
+		filters.AddFilter(neofs.DefaultStateAttribute, fmt.Sprintf("%d", startIndex*core.DefaultStateSyncInterval), object.MatchStringEqual)
+		prm.SetFilters(filters)
+
+		ctx, cancel := context.WithTimeout(sfs.basic.Ctx, sfs.cfg.Timeout)
+		var (
+			blockOidsObject []oid.ID
+			err             error
+		)
+		err = sfs.basic.Retry(func() error {
+			blockOidsObject, err = neofs.ObjectSearch(ctx, sfs.basic.Pool, sfs.basic.Account.PrivateKey(), sfs.cfg.ContainerID, prm)
+			return err
+		})
+		cancel()
+		if err != nil {
+			sfs.isShutdown.CompareAndSwap(false, true)
+			return 0, fmt.Errorf("failed to find '%s' object with index %d: %w", neofs.DefaultStateAttribute, startIndex, err)
+		}
+		if len(blockOidsObject) == 0 {
+			sfs.lastStateIndex = (startIndex - 1) * core.DefaultStateSyncInterval
+			return sfs.lastStateIndex, nil
+		}
+		startIndex++
+	}
+}
+
+// Start begins the state fetching and MPT update process.
+func (sfs *Service) Start() error {
+	if sfs.IsShutdown() {
+		return errors.New("service is already shut down")
+	}
+	if !sfs.isActive.CompareAndSwap(false, true) {
+		return nil
+	}
+	sfs.log.Info("NeoFS state fetcher service started")
+	sfs.wg.Add(1)
+	go sfs.run()
+	return nil
+}
+
+// Shutdown stops the state fetching process and persists the current state.
+func (sfs *Service) Shutdown() {
+	if !sfs.isActive.CompareAndSwap(true, false) {
+		return
+	}
+	sfs.basic.CtxCancel()
+	sfs.wg.Wait()
+
+	sfs.isShutdown.Store(true)
+	if sfs.shutdownCallback != nil {
+		sfs.shutdownCallback()
+	}
+	sfs.log.Info("NeoFS state fetcher service shutdown completed")
+}
+
+func (sfs *Service) run() {
+	var (
+		currentHeaderHeight = sfs.chain.HeaderHeight()
+		storedRoot          *state.MPTRoot
+		storedHeight        uint32
+		localMPT            *mpt.Trie
+		syncHeight          uint32
+		batch               = make(map[string][]byte, defaultBatchSize)
+		processed           = 0
+		neofsRoot           util.Uint256
+		lastKey             []byte
+	)
+
+	defer func() {
+		if localMPT != nil && len(lastKey) > 0 {
+			sfs.log.Info("saving checkpoint on shutdown",
+				zap.String("root", localMPT.StateRoot().String()),
+				zap.Uint32("height", syncHeight))
+			if len(batch) > 0 {
+				if err := sfs.chain.AddMPTBatch(batch, syncHeight); err != nil {
+					sfs.log.Error("failed to process final batch during shutdown", zap.Error(err))
+				}
+				if _, err := localMPT.PutBatch(mpt.MapToMPTBatch(batch)); err != nil {
+					sfs.log.Error("failed to update local MPT during shutdown", zap.Error(err))
+				}
+				localMPT.Flush(syncHeight)
+			}
+			tempState := checkpointState{
+				SyncHeight: syncHeight,
+				MPTRoot:    localMPT.StateRoot(),
+				NeoFSRoot:  neofsRoot,
+				LastKey:    lastKey,
+			}
+			if err := sfs.saveCheckpointState(tempState); err != nil {
+				sfs.log.Error("failed to save state checkpoint", zap.Error(err))
+			}
```

Comment:
This part should be managed by statesync module (see the related reviewcomments in statesync module file). StateFetcher should only manage storage items downloading and should then pass these raw storage items to statesync module (exactly like P2P layer passes raw MPT data to statesync Module).